### ISSUE TITLE
cannot import name '_PUBLIC_KEY_TYPES' from 'cryptography.x509.base' (/opt/anaconda3/lib/python3.8/site-packages/cryptography/x509/base.py)

### ISSUE THREAD

**Issue by mihirahuja1** [pyca/cryptography#8772]
Encountering this error while importing snowflake.connector package

pyOpenSSL==16.2.0
cryptography==3.4.8
snowflake-connector-python ==2.6.0
snowflake-sqlalchemy ==1.4.4

**Comment by alex** [comment]
Can you provide a full backtrace? It's a bit hard to debug with limited information.

Please note that cryptography 3.4.8 is quite old, and your best bet is likely to upgrade.

**Comment by mihirahuja1** [comment]
I tried upgrading to one of the latest versions, but since I'm using snowflake connector, it prompts me - snowflake connector is compatible with cryptography < 4.0.0. 

ImportError                               Traceback (most recent call last)
<ipython-input-16-cfffcd32aac7> in <module>
----> 1 import snowflake.connector
      2 

/opt/anaconda3/lib/python3.8/site-packages/snowflake/connector/__init__.py in <module>
     13 from logging import NullHandler
     14 
---> 15 from .connection import SnowflakeConnection
     16 from .cursor import DictCursor
     17 from .dbapi import (

/opt/anaconda3/lib/python3.8/site-packages/snowflake/connector/connection.py in <module>
     33 
     34 from . import errors, proxy
---> 35 from .auth import Auth
     36 from .auth_default import AuthByDefault
     37 from .auth_idtoken import AuthByIdToken

/opt/anaconda3/lib/python3.8/site-packages/snowflake/connector/auth.py in <module>
     27 )
     28 
---> 29 from .auth_keypair import AuthByKeyPair
     30 from .auth_usrpwdmfa import AuthByUsrPwdMfa
     31 from .compat import IS_LINUX, IS_MACOS, IS_WINDOWS, urlencode

/opt/anaconda3/lib/python3.8/site-packages/snowflake/connector/auth_keypair.py in <module>
     23 from .errorcode import ER_INVALID_PRIVATE_KEY
     24 from .errors import ProgrammingError
---> 25 from .network import KEY_PAIR_AUTHENTICATOR
     26 
     27 logger = getLogger(__name__)

/opt/anaconda3/lib/python3.8/site-packages/snowflake/connector/network.py in <module>
     20 import OpenSSL.SSL
     21 
---> 22 from . import ssl_wrap_socket
     23 from .compat import (
     24     BAD_GATEWAY,

/opt/anaconda3/lib/python3.8/site-packages/snowflake/connector/ssl_wrap_socket.py in <module>
     26 from .errors import OperationalError
     27 from .vendored.urllib3 import connection as connection_
---> 28 from .vendored.urllib3.contrib.pyopenssl import PyOpenSSLContext
     29 from .vendored.urllib3.util import ssl_ as ssl_
     30 

/opt/anaconda3/lib/python3.8/site-packages/snowflake/connector/vendored/urllib3/contrib/pyopenssl.py in <module>
     51 from cryptography import x509
     52 from cryptography.hazmat.backends.openssl import backend as openssl_backend
---> 53 from cryptography.hazmat.backends.openssl.x509 import _Certificate
     54 
     55 try:

/opt/anaconda3/lib/python3.8/site-packages/cryptography/hazmat/backends/openssl/x509.py in <module>
     23 )
     24 from cryptography.hazmat.primitives import hashes, serialization
---> 25 from cryptography.x509.base import _PUBLIC_KEY_TYPES
     26 from cryptography.x509.name import _ASN1Type
     27 

ImportError: cannot import name '_PUBLIC_KEY_TYPES' from 'cryptography.x509.base' (/opt/anaconda3/lib/python3.8/site-packages/cryptography/x509/base.py)

**Comment by mihirahuja1** [comment]
> Can you provide a full backtrace? It's a bit hard to debug with limited information.
> 
> Please note that cryptography 3.4.8 is quite old, and your best bet is likely to upgrade.

I tried upgrading to one of the latest versions, but since I'm using snowflake connector, it prompts me - snowflake connector is compatible with cryptography < 4.0.0.

ImportError Traceback (most recent call last)
in
----> 1 import snowflake.connector
2

/opt/anaconda3/lib/python3.8/site-packages/snowflake/connector/init.py in
13 from logging import NullHandler
14
---> 15 from .connection import SnowflakeConnection
16 from .cursor import DictCursor
17 from .dbapi import (

/opt/anaconda3/lib/python3.8/site-packages/snowflake/connector/connection.py in
33
34 from . import errors, proxy
---> 35 from .auth import Auth
36 from .auth_default import AuthByDefault
37 from .auth_idtoken import AuthByIdToken

/opt/anaconda3/lib/python3.8/site-packages/snowflake/connector/auth.py in
27 )
28
---> 29 from .auth_keypair import AuthByKeyPair
30 from .auth_usrpwdmfa import AuthByUsrPwdMfa
31 from .compat import IS_LINUX, IS_MACOS, IS_WINDOWS, urlencode

/opt/anaconda3/lib/python3.8/site-packages/snowflake/connector/auth_keypair.py in
23 from .errorcode import ER_INVALID_PRIVATE_KEY
24 from .errors import ProgrammingError
---> 25 from .network import KEY_PAIR_AUTHENTICATOR
26
27 logger = getLogger(name)

/opt/anaconda3/lib/python3.8/site-packages/snowflake/connector/network.py in
20 import OpenSSL.SSL
21
---> 22 from . import ssl_wrap_socket
23 from .compat import (
24 BAD_GATEWAY,

/opt/anaconda3/lib/python3.8/site-packages/snowflake/connector/ssl_wrap_socket.py in
26 from .errors import OperationalError
27 from .vendored.urllib3 import connection as connection_
---> 28 from .vendored.urllib3.contrib.pyopenssl import PyOpenSSLContext
29 from .vendored.urllib3.util import ssl_ as ssl_
30

/opt/anaconda3/lib/python3.8/site-packages/snowflake/connector/vendored/urllib3/contrib/pyopenssl.py in
51 from cryptography import x509
52 from cryptography.hazmat.backends.openssl import backend as openssl_backend
---> 53 from cryptography.hazmat.backends.openssl.x509 import _Certificate
54
55 try:

/opt/anaconda3/lib/python3.8/site-packages/cryptography/hazmat/backends/openssl/x509.py in
23 )
24 from cryptography.hazmat.primitives import hashes, serialization
---> 25 from cryptography.x509.base import _PUBLIC_KEY_TYPES
26 from cryptography.x509.name import _ASN1Type
27

ImportError: cannot import name '_PUBLIC_KEY_TYPES' from 'cryptography.x509.base' (/opt/anaconda3/lib/python3.8/site-packages/cryptography/x509/base.py)

**Comment by reaperhulk** [comment]
This error is likely due to a mixture of versions in the environment. Could you provide us instructions to minimally reproduce with a virtual environment? Upgrading to a newer snowflake-connector-python also allows newer cryptography versions, which may resolve this.

**Comment by mihirahuja1** [comment]
> This error is likely due to a mixture of versions in the environment. Could you provide us instructions to minimally reproduce with a virtual environment? Upgrading to a newer snowflake-connector-python also allows newer cryptography versions, which may resolve this.

I was using the following version - 
pyOpenSSL==16.2.0
cryptography==3.4.8
snowflake-connector-python ==2.6.0
snowflake-sqlalchemy ==1.4.4

It's interesting since I didn't encounter this issue up until recently.


I upgraded snowflake-connector-python and hence cryptography to newer versions. I'm faced with a different error now.

ImportError                               Traceback (most recent call last)
<ipython-input-18-cfffcd32aac7> in <module>
----> 1 import snowflake.connector
      2 

/opt/anaconda3/lib/python3.8/site-packages/snowflake/connector/__init__.py in <module>
     15 from logging import NullHandler
     16 
---> 17 from .connection import SnowflakeConnection
     18 from .cursor import DictCursor
     19 from .dbapi import (

/opt/anaconda3/lib/python3.8/site-packages/snowflake/connector/connection.py in <module>
     26 
     27 from . import errors, proxy
---> 28 from .auth import (
     29     FIRST_PARTY_AUTHENTICATORS,
     30     Auth,

/opt/anaconda3/lib/python3.8/site-packages/snowflake/connector/auth/__init__.py in <module>
      5 from __future__ import annotations
      6 
----> 7 from ._auth import Auth, get_public_key_fingerprint, get_token_from_private_key
      8 from .by_plugin import AuthByPlugin, AuthType
      9 from .default import AuthByDefault

/opt/anaconda3/lib/python3.8/site-packages/snowflake/connector/auth/_auth.py in <module>
     28 
     29 from ..compat import IS_LINUX, IS_MACOS, IS_WINDOWS, urlencode
---> 30 from ..constants import (
     31     DAY_IN_SECONDS,
     32     HTTP_HEADER_ACCEPT,

ImportError: cannot import name 'DAY_IN_SECONDS' from 'snowflake.connector.constants' (/opt/anaconda3/lib/python3.8/site-packages/snowflake/connector/constants.py)

**Comment by alex** [comment]
This error strongly suggests that your installation has become corrupted somehow. I strongly suggest you rebuild your environment from scratch.

**Comment by mihirahuja1** [comment]
> This error strongly suggests that your installation has become corrupted somehow. I strongly suggest you rebuild your environment from scratch.

Thanks for the suggestion. I tried rebuilding a fresh environment. 

Faced with a different error.

ImportError                               Traceback (most recent call last)
<ipython-input-4-cfffcd32aac7> in <module>
----> 1 import snowflake.connector
      2 

/opt/anaconda3/lib/python3.8/site-packages/snowflake/connector/__init__.py in <module>
     15 from logging import NullHandler
     16 
---> 17 from .connection import SnowflakeConnection
     18 from .cursor import DictCursor
     19 from .dbapi import (

/opt/anaconda3/lib/python3.8/site-packages/snowflake/connector/connection.py in <module>
     26 
     27 from . import errors, proxy
---> 28 from .auth import (
     29     FIRST_PARTY_AUTHENTICATORS,
     30     Auth,

/opt/anaconda3/lib/python3.8/site-packages/snowflake/connector/auth/__init__.py in <module>
      5 from __future__ import annotations
      6 
----> 7 from ._auth import Auth, get_public_key_fingerprint, get_token_from_private_key
      8 from .by_plugin import AuthByPlugin, AuthType
      9 from .default import AuthByDefault

/opt/anaconda3/lib/python3.8/site-packages/snowflake/connector/auth/_auth.py in <module>
     53     ServiceUnavailableError,
     54 )
---> 55 from ..network import (
     56     ACCEPT_TYPE_APPLICATION_SNOWFLAKE,
     57     CONTENT_TYPE_APPLICATION_JSON,

/opt/anaconda3/lib/python3.8/site-packages/snowflake/connector/network.py in <module>
     19 from typing import TYPE_CHECKING, Any
     20 
---> 21 import OpenSSL.SSL
     22 
     23 from snowflake.connector.secret_detector import SecretDetector

/opt/anaconda3/lib/python3.8/site-packages/OpenSSL/__init__.py in <module>
      6 """
      7 
----> 8 from OpenSSL import SSL, crypto
      9 from OpenSSL.version import (
     10     __author__,

/opt/anaconda3/lib/python3.8/site-packages/OpenSSL/SSL.py in <module>
      7 from weakref import WeakValueDictionary
      8 
----> 9 from OpenSSL._util import (
     10     UNSPECIFIED as _UNSPECIFIED,
     11     exception_from_error_queue as _exception_from_error_queue,

ImportError: cannot import name 'no_zero_allocator' from 'OpenSSL._util' (/opt/anaconda3/lib/python3.8/site-packages/OpenSSL/_util.py)

**Comment by reaperhulk** [comment]
That's interesting -- the name it can't import has been present for 7 years, so something very odd is going on here. Can you provide a method where we can reproduce this?

**Comment by mihirahuja1** [comment]
> That's interesting -- the name it can't import has been present for 7 years, so something very odd is going on here. Can you provide a method where we can reproduce this?

Indeed, something does look a bit off. Can you provide a bit more detail on how to proceed with a method to reproduce, sorry, I'm new to this coming form a research environment, so might be slow to understand this. 

Just an update, I'm able to easily call import snowflake.connector from python when opened in terminal, but the problem seems to be when I'm on jupyter notebook.

Also, Just downloaded Jupyter Lab for Desktop, started a new env, working. 

Maybe this detail could help.

**Comment by reaperhulk** [comment]
When I say a method to reproduce, I mean a set of steps that we can follow in our own environments that result in the same error occurring. To do this you'll want to find a way to minimize the issue, that is, find the minimum number of steps requires and remove all unrelated pieces. A common minimal reproducer would look something like:
* Create a new virtual environment using Python 3.11
* Run `pip install package==version package2==version`
* Run the following python code, which causes the exception: ```some block of code```

This can be a bit challenging, since sometimes the issue is actually a problem with a specific kind of Python installation (e.g. anaconda) or state living inside a jupyter notebook that comes with some larger package.

**Comment by mihirahuja1** [comment]
I did another attempt at creating a new virtual environment. Now it seems the code is working fine on a jupyter notebook as well. Can't really trace back what I did differently but here are the package versions I'm running.
I believe, the resolution was a combination of starting from scratch in a new environment and carefully managing the right package versioning.
numba                      0.56.4
numpy                      1.23.5
pyOpenSSL                  23.1.1
snowflake-connector-python 3.0.3
SQLAlchemy                 2.0.9
cryptography               40.0.2
